src/api/app/models/workflow/step/set_flags.rb

```diff
@@ -33,9 +34,12 @@ def set_flags
           existing_flag.update!(status: flag[:status])
         else
           main_object.add_flag(flag[:type], flag[:status], flag[:repository], flag[:architecture])
+          main_object.save!
         end
-        main_object.store(comment: 'SCM/CI integration, set_flags step')
+        objects_to_store << main_object
       end
+
+      objects_to_store.each { |main_object| main_object.store(comment: 'SCM/CI integration, set_flags step') }
```

Comment:
I think this will introduce a gap between the objects saved in [line 37](https://github.com/openSUSE/open-build-service/pull/15978/files#diff-0e61d122fd167de5a5db7074f1c09717fea6a74443d6f8ab037353283953bdd9R37) and the ones stored in the backend here, if there is a problem when saving: you could save some users, fail to save the next and then fail to store any of them, even the ones that are already saved.

Reply:
Sure. But we're safe in this case because all this code is surrounded by a transaction, in line 22.

Reply:
Mm but that's bad as well, because then we're going to store objects in the backend that then are rolled back in our database...

Reply:
We end up having a gap the other way around

Reply:
Before, we could have a gap because we could first store objects in the backend and then find an error and perform a rollout.

From now on, first we update existings flags or save new flags in the frontend and if no errors are found (and therefore, rolled out), we store those flags into the backend.

There could be more gaps, but I think this solution improves the one proposed before.

Reply:
You are right, if there is an error during saving nothing is done. :+1: